### ISSUE TITLE
RFC: New deep loading API

### ISSUE THREAD

**Issue by gdorsi** [garden-co/classic-jazz#1007]
**The problem**

Yesterday I've built a new example app to play around with the `writeOnly` role and got my deal of fighting with the deep load definitions.

Typescript doesn't help much on getting them right and any mistakes break the return type of the hooks.

So I've been thinking for a way to improve it and come up with this sketch:

```
  const { me } = useAccount(
    MusicaAccount.resolve({
      root: MusicaRoot.resolve({
        rootPlaylist: true,
        playlists: true,
      }),
    }),
  );
```

The advantages would be:

* Better Typescript integration
* Support for SchemaUnion (we can pass a callback to the resolve)
* Self-documented syntax, no more guessing
* We can actually make it backward compatible with the current system
* The declaration can be moved outside of the component:

```
const ResolveAccountWithPlaylists = MusicaAccount.resolve({
  root: MusicaRoot.resolve({
    rootPlaylist: true,
    playlists: true,
  }),
});

function Component() {
  const { me } = useAccount(ResolveAccountWithPlaylists)
```

* We can reduce the `useCoState` params to two:

```typescript
const root = useCoState(id, ResolveAccountRootWithPlaylists)
```

* The type definitions can be derived easily and distributed across the app:

```typescript
type ResolvedAccountWithPlaylists = Resolved<typeof ResolveAccountWithPlaylists>
```

**Comment by aeplay** [comment]
You are right that the current implementation of loading depths is very hard to get right / guess.<br><br>What you're suggesting seems a bit verbose and particularly since we want to move away from auto-loading and you'll have to write a lot more loading specs I want to keep it efficient to type.<br><br>Two simple things could make it a lot better:

1. Wrapping everything in a `{resolve: _}` like

```typescript
const personWithCompanyAndContact = useCoState(Person, personID, { resolve: { company: { mainContact: true } })
```

so even if you've never heard of loading depths is "reads" naturally

2. Changing the syntax for lists for consistency, so instead of

```typescript
const personWithChildrensTeachers = useCoState(Person, personID, { resolve: { children: [{ teacher: true }] })
```

we do

```typescript
const personWithChildrensTeachers = useCoState(Person, personID, { resolve: { children: { items: { teacher: true }}})
```

which most importantly should make the inference and auto-complete of the depth param much more robust!!<br><br>The latter would also allow introducing extra loading options that are particularly important for lists like pagination:

```typescript
const personWithChildrensTeachers = useCoState(Person, personID, { resolve: { children: { items: { teacher: true }, after: 100, limit: 10 }})
```

**Comment by aeplay** [comment]
I'll have a think about how I would do SchemaUnion in my suggested syntax, but probably also a callback
